presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -756,6 +854,16 @@ public static boolean isTypeOnlyCoercion(Type actualType, Type expectedType)
             return ((VarcharType) actualType).getLength().orElse(Integer.MAX_VALUE) < ((VarcharType) expectedType).getLength().orElse(Integer.MAX_VALUE);
         }
 
+        if (actualType instanceof DecimalType && expectedType instanceof DecimalType) {
+            DecimalType actualDecimal = (DecimalType) actualType;
+            DecimalType expectedDecimal = (DecimalType) expectedType;
+
+            if (actualDecimal.isShort() ^ expectedDecimal.isShort()) {
```

Comment:
We cannot type-only coerce decimals represented as Slice with those represented as long without